File: crates/prolly/bindings/ruby/examples/basic_map.rb
# frozen_string_literal: true

require 'prolly'

engine = Prolly::ProllyEngine.memory(Prolly.default_config)
tree = engine.create
tree = engine.put(tree, 'user:001'.b, 'Ada'.b)
tree = engine.put(tree, 'user:002'.b, 'Grace'.b)
tree = engine.put(tree, 'user:003'.b, 'Linus'.b)

raise 'expected user:001' unless engine.get(tree, 'user:001'.b) == 'Ada'.b

tree = engine.delete(tree, 'user:003'.b)
raise 'expected deleted user:003' unless engine.get(tree, 'user:003'.b).nil?

users = engine.range(tree, 'user:'.b, 'user;'.b)
expected = [['user:001'.b, 'Ada'.b], ['user:002'.b, 'Grace'.b]]
actual = users.map { |entry| [entry.key, entry.value] }
raise 'unexpected users' unless actual == expected

puts "basic_map: #{users.length} users in range"

File: crates/prolly/bindings/python/examples/basic_map.py
from __future__ import annotations

import prolly


engine = prolly.ProllyEngine.memory(prolly.default_config())

tree = engine.create()
tree = engine.put(tree, b"user:001", b"Ada")
tree = engine.put(tree, b"user:002", b"Grace")
tree = engine.put(tree, b"user:003", b"Linus")

assert engine.get(tree, b"user:001") == b"Ada"

tree = engine.delete(tree, b"user:003")
assert engine.get(tree, b"user:003") is None

users = engine.range(tree, b"user:", b"user;")
assert [(entry.key, entry.value) for entry in users] == [
    (b"user:001", b"Ada"),
    (b"user:002", b"Grace"),
]

print(f"basic_map: {len(users)} users in range")
